status-network-contracts/certora/specs/StakeVault.spec

```diff
@@ -1,14 +1,38 @@
+using ERC20A as staked;
+using StakeManager as stakeManager;
+
 methods {
   function owner() external returns (address) envfree;
   function isInitializing() external returns (bool) envfree;
-  // function ERC20A.balanceOf(address) external returns (uint256) envfree;
+  function depositedBalance() external returns (uint256) envfree;
+  function lockUntil() external returns (uint256) envfree;
+  function ERC20A.balanceOf(address) external returns (uint256) envfree;
   // function ERC20A.allowance(address, address) external returns(uint256) envfree;
-  // function ERC20A.totalSupply() external returns(uint256) envfree;
+  function ERC20A.totalSupply() external returns(uint256) envfree;
   // // function StakeManager.accounts(address) external returns(uint256, uint256, uint256, uint256, uint256, uint256) envfree;
   // function _.owner() external => DISPATCHER(true);
-  // function _.transfer(address, uint256) external => DISPATCHER(true);
+  function _.transfer(address, uint256) external => DISPATCHER(true);
+  function _.unstake(uint256) external => DISPATCHER(true);
 }
 
+invariant depositedBalanceLessEqualToERC20Balance()
+  depositedBalance() <= staked.balanceOf(currentContract)
+  filtered {
+      f -> f.contract == currentContract &&
+        f.selector != sig:migrateFromVault(uint256, uint256).selector
+  }
+  { preserved with (env e) {
+      require e.msg.sender != currentContract;
+      require staked.balanceOf(currentContract) + staked.balanceOf(e.msg.sender) <= to_mathint(staked.totalSupply());
+    }
+    preserved stake(uint256 amount, uint256 duration, address from) with (env e) {
+      require e.msg.sender != currentContract;
+      require from != currentContract;
+      require staked.balanceOf(currentContract) + staked.balanceOf(from) <= to_mathint(staked.totalSupply());
+    }
+  }
```

Comment:
@gravityblast here's a new rule I've added that ensure `depositedBalance()` is always `<= totalSupply()`

This actually uncovered a bug where `migratToVault()` would not reset `depositedBalance()` accordingly.
In fact, we're also not resetting properly the `lockUntil` field, which means, one could migrate back to a vault that already has a `lockUntil > 0`

Sending a PR for that too